code/modules/urist/modules/shipbattles/overmapships/components/component_weapons.dm

```diff
@@ -55,83 +59,126 @@
 ///datum/shipcomponents/weapons/proc/announce_fire()
 //	GLOB.global_announcer.autosay("<b>The attacking [mastership.ship_category] has fired a [src.name] at the ICS Nerva. Brace for impact.</b>", "[GLOB.using_map.full_name] Automated Defence Computer", "Common")
 
+/datum/shipcomponents/weapons/proc/locate_target(var/target_z)
+	var/target_x = rand(mastership.target_ship.target_x_bounds[1],mastership.target_ship.target_x_bounds[2])
+	var/target_y = rand(mastership.target_ship.target_y_bounds[1],mastership.target_ship.target_y_bounds[2])
+
+	var/turf/target = locate(target_x, target_y, target_z)
+
+	return target
+
+/datum/shipcomponents/weapons/proc/launch_projectile(var/turf/start_turf, var/turf/target_turf)
+	var/atom/movable/projectile = new projectile_type(start_turf)
+	if(istype(projectile, /obj/item/projectile))
+		var/obj/item/projectile/P = projectile
+		P.launch(target_turf) //projectiles have their own special proc
+
+	else if(istype(projectile, /obj/effect/meteor))
+		var/obj/effect/meteor/M = projectile
+		M.dest = target_turf
+		spawn(0)
+			walk_towards(M, M.dest, 3) //meteors do their own thing too
+
+	else
+		projectile.throw_at(target_turf) //anything else just uses the default throw proc. this potentially allows for ship weapons that do things like throw mobs. clown cannon anyone?
```

Comment:
Seems to me there's a general-purpose `launch_anything(var/turf/start_turf, var/turf/target_turf, var/projectile_type)` proc underneath the ship bits here that could be used for other stuff and the shipweapons could just feed it their projectile type and endpoints.